### ISSUE TITLE
Management of expired sponsorships

### ISSUE THREAD

**Issue by jrosindell** [OneZoom/OZtree#362]
Option 1. Cron job to move them into expired reservations table
Option 2. Make APIs look in both tables and account for whether species are active or not in their returns - then make the action of clicking to sponsor something that's expired be the action that moves it to the expired reservations table.

**Comment by jrosindell** [comment]
@lentinj if you have time in the upcoming sprint please can you write the scripts needed for these cron jobs?

**Comment by lentinj** [comment]
The web2py-way of doing this would probably be the [scheduler](http://www.web2py.com/books/default/chapter/29/04/the-core#web2py-Scheduler), a task-queue setup for managing workers (python process) that can run jobs. However, it feels fiddly to setup (you'd have to start/stop a separate python process, and [have something to kickstart the jobs](http://www.web2py.com/books/default/chapter/29/13#Initial-Queuing-of-a-Repeating-Task). Unless the reporting is something you really like, I think this isn't worth it.

Running a ``private/background_tasks.py`` script periodically would involve less moving parts, I'll put together an example of doing that. @hyanwong do you still use supervisord to manage the OneZoom processes?

**Comment by hyanwong** [comment]
I agree, a periodic script is better. I don't want to tie us in to the web2py ecosystem too hard.

Yes we use supervisord on the server.

**Comment by lentinj** [comment]
@hyanwong The above adds a ``private/background_tasks.py`` that can be added to a supervisord config. I'm assuming you can guess what to add to ``README_SERVER.markdown`` easily enough based on the example at the top. If you want me to check then can do.

Make sure it autorestarts though, since errors in the script aren't handled. I figured if it is failing it would make more noise this way, and any future runs will have a cleaner environment.

**Comment by hyanwong** [comment]
Perfect, thanks.

**Comment by lentinj** [comment]
Are OTTs flitting in and out of existence a likely scenario? I don't think renewing it for free is the right approach due to the database spew it will generate, adding new renewal rows every 90 days if an OTT permanently disappears.

Obviously asking someone to renew a non-existent OTT isn't good, but if it only stopped existing for a week, then they'd still get an e-mail after it reappeared. Sure the OTT may have expired by that point, but it's not like anyone could have stolen it for themselves in the interim.

The remaining task here is the reminder e-mail background job. Would it be okay to move the discussion to #430 and I take over that ticket? It'll help keep unrelated conversations separate I think.

@jrosindell reminder email questions:
* We have ``emailed_re_renewal_initial`` and ``emailed_re_renewal_final`` date fields in the DB. I guess there's a 2 strikes policy of "expiring soon" and "expiring really soon". Care to put numbers on those? "soon" == 90 days, "really soon" == a reminder 14 days after sending the previous?
* Presumably we send one-email with a list of species that are expiring, rather than an e-mail per species?
* Is sending separate initial and final e-mails at the same time okay, in the (astronomically unlikely) case that a user has both sponsorships that we e-mailed once about and newly expiring species.
* Do we have templates for what the initial / final e-mails will have in them?

**Comment by lentinj** [comment]
I've moved conversation about renewal reminder emails to #439. Leaving this open as a reminder that ``private/background_tasks.py`` needs to be added to the production supervisord config.

**Comment by lentinj** [comment]
The above adds a grunt task runner for the script, so you can do: ``grunt exec:background_tasks:--single:--verbose`` without worrying about web2py incantations.
